clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -3158,6 +3292,29 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
     int64_t N = Index.getExtValue();
     return Ts.getPackAsArray()[N].getAsType();
   }
+
+  case BTK__builtin_common_type: {
+    assert(Converted.size() == 4);
+    if (llvm::any_of(Converted, [](auto &C) { return C.isDependent(); }))
+      return Context.getCanonicalTemplateSpecializationType(TemplateName(BTD),
+                                                            Converted);
+
+    TemplateName BaseTemplate = Converted[0].getAsTemplate();
+    TemplateName HasTypeMember = Converted[1].getAsTemplate();
+    QualType HasNoTypeMember = Converted[2].getAsType();
+    ArrayRef<TemplateArgument> Ts = Converted[3].getPackAsArray();
+    if (auto CT = builtinCommonTypeImpl(SemaRef, BaseTemplate, TemplateLoc, Ts);
+        !CT.isNull()) {
+      TemplateArgumentListInfo TAs;
+      TAs.addArgument(TemplateArgumentLoc(
+          TemplateArgument(CT), SemaRef.Context.getTrivialTypeSourceInfo(
+                                    CT, TemplateArgs[1].getLocation())));
+
+      return SemaRef.CheckTemplateIdType(HasTypeMember, TemplateLoc, TAs);
+    }
+    return HasNoTypeMember;
```

Comment:
This is missing wrapping the result in a TemplateSpecializationType, just like the other builtin templates do, because users of CheckTemplateId assume that it returns a TST of whatever TemplateName it is checking.
Otherwise, these users will wrap the resulting type in a TemplateSpecializationTypeLoc which is unrelated to what is returned.

Worse, in the negative case, HasNoTypeMember can be something which is not a TST at all, and this will lead to crashes when we try to build a TypeLoc with the wrong kind.

Reply:
I'm not sure I understand. `__type_pack_element` above also just returns a type in the usual case.